presto-main/src/main/java/com/facebook/presto/sql/planner/plan/WindowNode.java

```diff
@@ -67,15 +70,36 @@ public WindowNode(
         checkArgument(specification.getPartitionBy().containsAll(prePartitionedInputs), "prePartitionedInputs must be contained in partitionBy");
         checkArgument(preSortedOrderPrefix <= specification.getOrderBy().size(), "Cannot have sorted more symbols than those requested");
         checkArgument(preSortedOrderPrefix == 0 || ImmutableSet.copyOf(prePartitionedInputs).equals(ImmutableSet.copyOf(specification.getPartitionBy())), "preSortedOrderPrefix can only be greater than zero if all partition symbols are pre-partitioned");
+        checkArgument(allFunctionsHaveFrames(windowFunctions, frames), "not all functions have frames assigned");
 
         this.source = source;
         this.prePartitionedInputs = ImmutableSet.copyOf(prePartitionedInputs);
         this.specification = specification;
         this.windowFunctions = ImmutableMap.copyOf(windowFunctions);
+        this.frames = ImmutableMap.copyOf(frames);
         this.hashSymbol = hashSymbol;
         this.preSortedOrderPrefix = preSortedOrderPrefix;
     }
 
+    @JsonCreator
+    public WindowNode(
+            @JsonProperty("id") PlanNodeId id,
+            @JsonProperty("source") PlanNode source,
+            @JsonProperty("specification") Specification specification,
+            @JsonProperty("windowFunctions") Map<Symbol, Function> windowFunctions,
+            @JsonProperty("frames") List<FunctionWithFrame> frames,
+            @JsonProperty("hashSymbol") Optional<Symbol> hashSymbol,
+            @JsonProperty("prePartitionedInputs") Set<Symbol> prePartitionedInputs,
+            @JsonProperty("preSortedOrderPrefix") int preSortedOrderPrefix)
+    {
+        this(id, source, specification, windowFunctions, frames.stream().collect(toImmutableMap(f -> f.getFunction(), f -> f.getFrame())), hashSymbol, prePartitionedInputs, preSortedOrderPrefix);
+    }
+
+    private static boolean allFunctionsHaveFrames(Map<Symbol, Function> windowFunctions, Map<Function, Frame> frames)
+    {
+        return windowFunctions.values().stream().filter(f -> frames.containsKey(f)).count() == windowFunctions.size();
```

Comment:
It's already rewritten by @ebd2 in following commits I've already merged into this PR.